Empty project after reopening (instead of last workspace)

### Check for existing issues

- [X] Completed

### Describe the bug / provide steps to reproduce it

When closing Zed and then reopening, sometimes the workspace won't be restored and an empty project will be opened instead.

Reproduction steps:

1. Open any project in Zed
2. Close the window via the title bar close button, Ctrl + Shift + W or similar.
3. Reopen Zed.
4. Find that the project is not restored, but an empty project is instead.

Interestingly, I found that the following setting fixes the issue:
```json
{
  "restore_on_startup": "last_workspace"
}
``` 

### Environment

Zed: v0.148.0 (Zed Nightly 3bd9a3f478337deb4d46a711d5a3d5c0e4627216)
OS: Linux Wayland fedora 40
Memory: 30.5 GiB
Architecture: x86_64
GPU: AMD Radeon RX 6700 XT (RADV NAVI22) || radv || Mesa 24.1.4

### If applicable, add mockups / screenshots to help explain present your vision of the feature

_No response_

### If applicable, attach your Zed.log file to this issue.

<details><summary>Zed.log</summary><pre>
2024-08-01T15:28:51.226722309-03:00 [INFO] ========== starting zed ==========
2024-08-01T15:28:51.251541149-03:00 [INFO] perform;
2024-08-01T15:28:51.251637249-03:00 [INFO] read_command;
2024-08-01T15:28:51.251707039-03:00 [INFO] Opening main db
2024-08-01T15:28:51.251771339-03:00 [INFO] read_command;
2024-08-01T15:28:51.251916879-03:00 [INFO] socket reader;
2024-08-01T15:28:51.252103879-03:00 [INFO] new;
2024-08-01T15:28:51.252543198-03:00 [INFO] keep_updated;
2024-08-01T15:28:51.254421907-03:00 [INFO] Using git binary path: None
2024-08-01T15:28:51.273404042-03:00 [ERROR] No such file or directory (os error 2) about ["/home/apricotbucket28/.config/github-copilot"]
2024-08-01T15:28:51.273924541-03:00 [INFO] extensions updated. loading 8, reloading 0, unloading 0
2024-08-01T15:28:51.27521294-03:00 [INFO] activate is not implemented on Linux, ignoring the call
2024-08-01T15:28:51.27610899-03:00 [INFO] Opening main db
2024-08-01T15:28:51.276300199-03:00 [INFO] perform;
2024-08-01T15:28:51.276385079-03:00 [INFO] read_command;
2024-08-01T15:28:51.276479229-03:00 [INFO] read_command;
2024-08-01T15:28:51.276644109-03:00 [INFO] socket reader;
2024-08-01T15:28:51.279476977-03:00 [INFO] new;
2024-08-01T15:28:51.279992887-03:00 [INFO] keep_updated;
2024-08-01T15:28:51.299564101-03:00 [INFO] Enabling Vulkan Portability
2024-08-01T15:28:51.299632041-03:00 [INFO] Enabling color space support
2024-08-01T15:28:51.319606785-03:00 [INFO] Adapter "AMD Radeon RX 6700 XT (RADV NAVI22)"
2024-08-01T15:28:51.319679605-03:00 [INFO] Ray tracing is supported
2024-08-01T15:28:51.323857292-03:00 [INFO] Using surface present mode MAILBOX
2024-08-01T15:28:51.324135162-03:00 [WARN] Unable to forbid exclusive full screen
2024-08-01T15:28:51.324779171-03:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-01T15:28:51.324854271-03:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-01T15:28:51.324920041-03:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2024-08-01T15:28:51.328881128-03:00 [INFO] Using GPU: GPUSpecs { is_software_emulated: false, device_name: "AMD Radeon RX 6700 XT (RADV NAVI22)", driver_name: "radv", driver_info: "Mesa 24.1.4" }
2024-08-01T15:28:51.329189538-03:00 [INFO] set status on client 0: Authenticating
2024-08-01T15:28:51.329816997-03:00 [INFO] perform;
2024-08-01T15:28:51.329920507-03:00 [INFO] read_command;
2024-08-01T15:28:51.329999587-03:00 [INFO] Opening main db
2024-08-01T15:28:51.330101167-03:00 [INFO] read_command;
2024-08-01T15:28:51.330213767-03:00 [INFO] socket reader;
2024-08-01T15:28:51.335527723-03:00 [INFO] new;
2024-08-01T15:28:51.335905192-03:00 [INFO] keep_updated;
2024-08-01T15:28:51.362796791-03:00 [INFO] set environment variables from shell:/usr/bin/zsh, path:/home/apricotbucket28/.local/bin:/usr/lib64/ccache:/home/apricotbucket28/.cargo/bin:/usr/local/bin:/usr/bin:/bin:/usr/local/sbin:/usr/sbin:/sbin:/home/apricotbucket28/Development/flutter/bin:/home/apricotbucket28/.cargo/bin:/home/apricotbucket28/.local/zed-nightly.app/bin
2024-08-01T15:28:51.370537205-03:00 [INFO] Using surface present mode MAILBOX
2024-08-01T15:28:51.370850615-03:00 [WARN] Unable to forbid exclusive full screen
2024-08-01T15:28:51.371419134-03:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: Ignored }
2024-08-01T15:28:51.374642832-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.375555961-03:00 [INFO] Using surface present mode MAILBOX
2024-08-01T15:28:51.375833171-03:00 [WARN] Unable to forbid exclusive full screen
2024-08-01T15:28:51.376066971-03:00 [INFO] Initializing Blade pipelines for surface SurfaceInfo { format: Bgra8UnormSrgb, alpha: PreMultiplied }
2024-08-01T15:28:51.378797759-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.378949888-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.379511608-03:00 [INFO] Using surface present mode MAILBOX
2024-08-01T15:28:51.379789978-03:00 [WARN] Unable to forbid exclusive full screen
2024-08-01T15:28:51.383804784-03:00 [INFO] set status on client 119342: Connecting
2024-08-01T15:28:51.384705734-03:00 [INFO] Opening main db
2024-08-01T15:28:51.390416909-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.392366048-03:00 [ERROR] no worktrees when constructing ProjectLspAdapterDelegate
2024-08-01T15:28:51.394555086-03:00 [INFO] perform;
2024-08-01T15:28:51.394661996-03:00 [INFO] read_command;
2024-08-01T15:28:51.394771806-03:00 [INFO] read_command;
2024-08-01T15:28:51.395027816-03:00 [INFO] socket reader;
2024-08-01T15:28:51.398391553-03:00 [INFO] perform;
2024-08-01T15:28:51.398489613-03:00 [INFO] read_command;
2024-08-01T15:28:51.398568293-03:00 [INFO] read_command;
2024-08-01T15:28:51.398651683-03:00 [INFO] socket reader;
2024-08-01T15:28:51.411690482-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.41419622-03:00 [INFO] perform;
2024-08-01T15:28:51.41430663-03:00 [INFO] read_command;
2024-08-01T15:28:51.41440514-03:00 [INFO] read_command;
2024-08-01T15:28:51.41454106-03:00 [INFO] socket reader;
2024-08-01T15:28:51.417332398-03:00 [INFO] perform;
2024-08-01T15:28:51.417398548-03:00 [INFO] read_command;
2024-08-01T15:28:51.417464618-03:00 [INFO] read_command;
2024-08-01T15:28:51.417540958-03:00 [INFO] socket reader;
2024-08-01T15:28:51.430816137-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.431426977-03:00 [INFO] perform;
2024-08-01T15:28:51.431534357-03:00 [INFO] read_command;
2024-08-01T15:28:51.431613477-03:00 [INFO] read_command;
2024-08-01T15:28:51.431727137-03:00 [INFO] socket reader;
2024-08-01T15:28:51.434549275-03:00 [INFO] perform;
2024-08-01T15:28:51.434633115-03:00 [INFO] read_command;
2024-08-01T15:28:51.434710714-03:00 [INFO] read_command;
2024-08-01T15:28:51.434809984-03:00 [INFO] socket reader;
2024-08-01T15:28:51.657733588-03:00 [INFO] Initializing default prettier with plugins {}
2024-08-01T15:28:51.658567368-03:00 [INFO] perform;
2024-08-01T15:28:51.658645268-03:00 [INFO] read_command;
2024-08-01T15:28:51.658709898-03:00 [INFO] read_command;
2024-08-01T15:28:51.658810938-03:00 [INFO] socket reader;
2024-08-01T15:28:51.663640704-03:00 [INFO] perform;
2024-08-01T15:28:51.663735514-03:00 [INFO] read_command;
2024-08-01T15:28:51.663816593-03:00 [INFO] read_command;
2024-08-01T15:28:51.663920333-03:00 [INFO] socket reader;
2024-08-01T15:28:51.66789427-03:00 [INFO] connected to rpc endpoint https://collab.zed.dev/rpc
2024-08-01T15:28:52.18728989-03:00 [INFO] add connection to peer
2024-08-01T15:28:52.18736361-03:00 [INFO] add_connection;
2024-08-01T15:28:52.18746374-03:00 [INFO] waiting for server hello
2024-08-01T15:28:52.215157128-03:00 [INFO] got server hello
2024-08-01T15:28:52.215224278-03:00 [INFO] set status to connected (connection id: ConnectionId { owner_id: 0, id: 0 }, peer id: PeerId { owner_id: 462, id: 862635 })
2024-08-01T15:28:52.215332758-03:00 [INFO] set status on client 119342: Connected { peer_id: PeerId { owner_id: 462, id: 862635 }, connection_id: ConnectionId { owner_id: 0, id: 0 } }
2024-08-01T15:28:54.103937297-03:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-01T15:28:54.115375358-03:00 [INFO] Creating a descriptor pool for at most 16 sets
2024-08-01T15:28:54.128757017-03:00 [INFO] Creating a descriptor pool for at most 256 sets
2024-08-01T15:28:54.142058817-03:00 [INFO] Creating a descriptor pool for at most 256 sets
</details>

--- a/crates/collab/src/tests/editor_tests.rs
+++ b/crates/collab/src/tests/editor_tests.rs
@@ -43,7 +43,7 @@ use std::{
     },
 };
 use text::Point;
-use workspace::Workspace;
+use workspace::{CloseIntent, Workspace};
 
 #[gpui::test(iterations = 10)]
 async fn test_host_disconnect(
@@ -134,7 +134,9 @@ async fn test_host_disconnect(
 
     // Ensure client B is not prompted to save edits when closing window after disconnecting.
     let can_close = workspace_b
-        .update(cx_b, |workspace, cx| workspace.prepare_to_close(true, cx))
+        .update(cx_b, |workspace, cx| {
+            workspace.prepare_to_close(CloseIntent::Quit, cx)
+        })
         .unwrap()
         .await
         .unwrap();

--- a/crates/recent_projects/src/recent_projects.rs
+++ b/crates/recent_projects/src/recent_projects.rs
@@ -32,7 +32,8 @@ use ui::{
 };
 use util::{paths::PathExt, ResultExt};
 use workspace::{
-    AppState, ModalView, SerializedWorkspaceLocation, Workspace, WorkspaceId, WORKSPACE_DB,
+    AppState, CloseIntent, ModalView, SerializedWorkspaceLocation, Workspace, WorkspaceId,
+    WORKSPACE_DB,
 };
 
 #[derive(PartialEq, Clone, Deserialize, Default)]
@@ -311,7 +312,7 @@ impl PickerDelegate for RecentProjectsDelegate {
                                     cx.spawn(move |workspace, mut cx| async move {
                                         let continue_replacing = workspace
                                             .update(&mut cx, |workspace, cx| {
-                                                workspace.prepare_to_close(true, cx)
+                                                workspace.prepare_to_close(CloseIntent::ReplaceWindow, cx)
                                             })?
                                             .await?;
                                         if continue_replacing {
@@ -570,7 +571,7 @@ fn open_dev_server_project(
             cx.spawn(move |workspace, mut cx| async move {
                 let continue_replacing = workspace
                     .update(&mut cx, |workspace, cx| {
-                        workspace.prepare_to_close(true, cx)
+                        workspace.prepare_to_close(CloseIntent::ReplaceWindow, cx)
                     })?
                     .await?;
                 if continue_replacing {

--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -224,6 +224,16 @@ impl_actions!(
     ]
 );
 
+#[derive(PartialEq, Eq, Debug)]
+pub enum CloseIntent {
+    /// Quit the program entirely.
+    Quit,
+    /// Close a window.
+    CloseWindow,
+    /// Replace the workspace in an existing window.
+    ReplaceWindow,
+}
+
 #[derive(Clone)]
 pub struct Toast {
     id: NotificationId,
@@ -1612,9 +1622,7 @@ impl Workspace {
     }
 
     pub fn close_window(&mut self, _: &CloseWindow, cx: &mut ViewContext<Self>) {
-        // On Linux and Windows, closing the last window also exits the entire program.
-        let quitting = cfg!(not(target_os = "macos")) && cx.windows().len() == 1;
-        let prepare = self.prepare_to_close(quitting, cx);
+        let prepare = self.prepare_to_close(CloseIntent::CloseWindow, cx);
         let window = cx.window_handle();
         cx.spawn(|_, mut cx| async move {
             if prepare.await? {
@@ -1629,11 +1637,16 @@ impl Workspace {
 
     pub fn prepare_to_close(
         &mut self,
-        quitting: bool,
+        close_intent: CloseIntent,
         cx: &mut ViewContext<Self>,
     ) -> Task<Result<bool>> {
         let active_call = self.active_call().cloned();
         let window = cx.window_handle();
+
+        // On Linux and Windows, closing the last window should restore the last workspace.
+        let save_last_workspace = cfg!(not(target_os = "macos"))
+            && close_intent != CloseIntent::ReplaceWindow
+            && cx.windows().len() == 1;
 
         cx.spawn(|this, mut cx| async move {
             let workspace_count = (*cx).update(|cx| {
@@ -1644,7 +1657,7 @@ impl Workspace {
             })?;
 
             if let Some(active_call) = active_call {
-                if !quitting
+                if close_intent != CloseIntent::Quit
                     && workspace_count == 1
                     && active_call.read_with(&cx, |call, _| call.room().is_some())?
                 {
@@ -1676,7 +1689,10 @@ impl Workspace {
 
             // If we're not quitting, but closing, we remove the workspace from
             // the current session.
-            if !quitting && save_result.as_ref().map_or(false, |&res| res) {
+            if close_intent != CloseIntent::Quit
+                && !save_last_workspace
+                && save_result.as_ref().map_or(false, |&res| res)
+            {
                 this.update(&mut cx, |this, cx| this.remove_from_session(cx))?
                     .await;
             }
@@ -5573,7 +5589,7 @@ pub fn reload(reload: &Reload, cx: &mut AppContext) {
         // If the user cancels any save prompt, then keep the app open.
         for window in workspace_windows {
             if let Ok(should_close) = window.update(&mut cx, |workspace, cx| {
-                workspace.prepare_to_close(true, cx)
+                workspace.prepare_to_close(CloseIntent::Quit, cx)
             }) {
                 if !should_close.await? {
                     return Ok(());
@@ -5777,7 +5793,7 @@ mod tests {
         workspace.update(cx, |w, cx| {
             w.add_item_to_active_pane(Box::new(item1.clone()), None, true, cx)
         });
-        let task = workspace.update(cx, |w, cx| w.prepare_to_close(false, cx));
+        let task = workspace.update(cx, |w, cx| w.prepare_to_close(CloseIntent::CloseWindow, cx));
         assert!(task.await.unwrap());
 
         // When there are dirty untitled items, prompt to save each one. If the user
@@ -5792,7 +5808,7 @@ mod tests {
             w.add_item_to_active_pane(Box::new(item2.clone()), None, true, cx);
             w.add_item_to_active_pane(Box::new(item3.clone()), None, true, cx);
         });
-        let task = workspace.update(cx, |w, cx| w.prepare_to_close(false, cx));
+        let task = workspace.update(cx, |w, cx| w.prepare_to_close(CloseIntent::CloseWindow, cx));
         cx.executor().run_until_parked();
         cx.simulate_prompt_answer(2); // cancel save all
         cx.executor().run_until_parked();
@@ -5833,7 +5849,7 @@ mod tests {
             w.add_item_to_active_pane(Box::new(item1.clone()), None, true, cx);
             w.add_item_to_active_pane(Box::new(item2.clone()), None, true, cx);
         });
-        let task = workspace.update(cx, |w, cx| w.prepare_to_close(false, cx));
+        let task = workspace.update(cx, |w, cx| w.prepare_to_close(CloseIntent::CloseWindow, cx));
         assert!(task.await.unwrap());
     }
 

--- a/crates/workspace/src/workspace.rs
+++ b/crates/workspace/src/workspace.rs
@@ -1612,8 +1612,10 @@ impl Workspace {
     }
 
     pub fn close_window(&mut self, _: &CloseWindow, cx: &mut ViewContext<Self>) {
+        // On Linux and Windows, closing the last window also exits the entire program.
+        let quitting = cfg!(not(target_os = "macos")) && cx.windows().len() == 1;
+        let prepare = self.prepare_to_close(quitting, cx);
         let window = cx.window_handle();
-        let prepare = self.prepare_to_close(false, cx);
         cx.spawn(|_, mut cx| async move {
             if prepare.await? {
                 window.update(&mut cx, |_, cx| {

--- a/crates/zed/src/zed.rs
+++ b/crates/zed/src/zed.rs
@@ -35,6 +35,7 @@ use std::{borrow::Cow, ops::Deref, path::Path, sync::Arc};
 use task::static_source::{StaticSource, TrackedFile};
 use theme::ActiveTheme;
 use workspace::notifications::NotificationId;
+use workspace::CloseIntent;
 
 use paths::{local_settings_file_relative_path, local_tasks_file_relative_path};
 use terminal_view::terminal_panel::{self, TerminalPanel};
@@ -615,7 +616,7 @@ fn quit(_: &Quit, cx: &mut AppContext) {
         for window in workspace_windows {
             if let Some(should_close) = window
                 .update(&mut cx, |workspace, cx| {
-                    workspace.prepare_to_close(true, cx)
+                    workspace.prepare_to_close(CloseIntent::Quit, cx)
                 })
                 .log_err()
             {

